File: python/terminai_agent/tools/grep_files.py
"""Tool for searching files with grep-like functionality."""

import logging
import re
from pathlib import Path

from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

# Maximum number of matches to return
MAX_MATCHES = 100

# Maximum number of files to search
MAX_FILES = 1000


class GrepFilesArgs(BaseModel):
    """Arguments for grep_files tool."""

    pattern: str = Field(description="Pattern to search for (regex or literal string)")
    file_pattern: str | None = Field(
        None, description="File glob pattern (e.g., '*.rs', 'src/**/*.py'). Optional."
    )
    case_insensitive: bool = Field(False, description="Whether to use case-insensitive search")
    max_matches: int | None = Field(
        None, description=f"Maximum number of matches to return (default: 50, max: {MAX_MATCHES})"
    )


class Match(BaseModel):
    """A single match in a file."""

    file: str = Field(description="Relative path to the file")
    line_number: int = Field(description="Line number (1-indexed)")
    line: str = Field(description="Matching line content")


class GrepFilesResult(BaseModel):
    """Result from grep_files tool."""

    matches: list[Match] = Field(description="List of matches found")
    pattern: str = Field(description="Pattern that was searched")
    num_files_searched: int = Field(description="Number of files that were searched")
    truncated: bool = Field(description="Whether results were truncated")
    error: str | None = Field(None, description="Error message if search failed")


def should_skip_file(path: Path) -> bool:
    """Determine if a file should be skipped during search.

    Args:
        path: Path to check

    Returns:
        True if file should be skipped, False otherwise
    """
    # Skip hidden files and directories
    if any(part.startswith(".") for part in path.parts):
        return True

    # Skip common directories
    skip_dirs = {"node_modules", "target", "__pycache__", ".git", ".venv", "venv"}
    if any(part in skip_dirs for part in path.parts):
        return True

    # Skip binary files by extension
    binary_extensions = {".pyc", ".so", ".dylib", ".dll", ".exe", ".bin", ".o"}
    if path.suffix in binary_extensions:
        return True

    # Skip large files (> 10MB)
    try:
        if path.stat().st_size > 10 * 1024 * 1024:
            return True
    except OSError:
        return True

    return False


async def grep_files(args: GrepFilesArgs, cwd: str) -> GrepFilesResult:
    """Search for a pattern in files under the current working directory.

    Args:
        args: Tool arguments
        cwd: Current working directory

    Returns:
        GrepFilesResult with matches or error
    """
    cwd_path = Path(cwd)
    matches: list[Match] = []
    files_searched = 0
    max_matches = min(args.max_matches or 50, MAX_MATCHES)

    # Compile regex pattern
    try:
        flags = re.IGNORECASE if args.case_insensitive else 0
        pattern = re.compile(args.pattern, flags)
    except re.error as e:
        return GrepFilesResult(
            matches=[],
            pattern=args.pattern,
            num_files_searched=0,
            truncated=False,
            error=f"Invalid regex pattern: {e}",
        )

    # Determine which files to search
    if args.file_pattern:
        # Use glob pattern
        try:
            file_paths = list(cwd_path.glob(args.file_pattern))
        except ValueError as e:
            return GrepFilesResult(
                matches=[],
                pattern=args.pattern,
                num_files_searched=0,
                truncated=False,
                error=f"Invalid file pattern: {e}",
            )
    else:
        # Search all files recursively
        file_paths = list(cwd_path.rglob("*"))

    # Search files
    for file_path in file_paths:
        if len(matches) >= max_matches or files_searched >= MAX_FILES:
            break

        # Skip if not a file or should be skipped
        if not file_path.is_file() or should_skip_file(file_path):
            continue

        files_searched += 1

        try:
            # Read file and search
            content = file_path.read_text()
            for line_num, line in enumerate(content.splitlines(), start=1):
                if pattern.search(line):
                    # Get relative path from cwd
                    try:
                        relative_path = file_path.relative_to(cwd_path)
                    except ValueError:
                        relative_path = file_path

                    matches.append(
                        Match(
                            file=str(relative_path),
                            line_number=line_num,
                            line=line.rstrip(),
                        )
                    )

                    if len(matches) >= max_matches:
                        break
        except (UnicodeDecodeError, PermissionError):
            # Skip files we can't read
            continue
        except Exception as e:
            logger.debug(f"Error searching file {file_path}: {e}")
            continue

    truncated = len(matches) >= max_matches or files_searched >= MAX_FILES

    return GrepFilesResult(
        matches=matches,
        pattern=args.pattern,
        num_files_searched=files_searched,
        truncated=truncated,
    )


def format_grep_result(result: GrepFilesResult) -> str:
    """Format grep result as a human-readable string.

    Args:
        result: GrepFilesResult to format

    Returns:
        Formatted string
    """
    if result.error:
        return f"Error: {result.error}"

    if not result.matches:
        return f"No matches found for pattern: {result.pattern}"

    output = [f"## Grep Results for '{result.pattern}'\n"]
    output.append(f"Found {len(result.matches)} matches:\n")

    current_file = None
    for match in result.matches:
        # Show file path when it changes
        if current_file != match.file:
            if current_file is not None:
                output.append("\n")
            output.append(f"### {match.file}\n\n")
            current_file = match.file

        output.append(f"{match.line_number}:  {match.line}\n")

    if result.truncated:
        output.append(f"\n*Showing first {len(result.matches)} matches. There may be more.*")

    return "".join(output)

File: src/llm_old/tools/grep_files.rs
use grep::matcher::Matcher;
use grep::regex::RegexMatcher;
use grep::searcher::{BinaryDetection, SearcherBuilder};
use rig::completion::ToolDefinition;
use rig::tool::Tool;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::path::{Path, PathBuf};
use std::sync::{Arc, RwLock};
use walkdir::WalkDir;

/// Maximum number of matches to return
const MAX_MATCHES: usize = 100;

/// Maximum number of files to search
const MAX_FILES: usize = 1000;

#[derive(Deserialize)]
pub struct GrepFilesArgs {
  /// Pattern to search for (regex or literal string)
  pattern: String,
  /// File glob pattern (e.g., "*.rs", "src/**/*.py")
  #[serde(default)]
  file_pattern: Option<String>,
  /// Whether to use case-insensitive search
  #[serde(default)]
  case_insensitive: bool,
  /// Maximum number of matches to return (default: 50, max: 100)
  #[serde(default)]
  max_matches: Option<usize>,
}

#[derive(Debug, thiserror::Error)]
pub enum GrepFilesError {
  #[error("Invalid regex pattern: {0}")]
  InvalidPattern(String),
  #[error("IO error: {0}")]
  IoError(#[from] std::io::Error),
  #[error("Walk directory error: {0}")]
  WalkDirError(#[from] walkdir::Error),
  #[error("Failed to acquire cwd lock")]
  LockError,
  #[error("Search error: {0}")]
  SearchError(String),
}

#[derive(Debug)]
struct Match {
  file: PathBuf,
  line_number: u64,
  line: String,
}

/// Tool for searching file contents using grep
pub struct GrepFilesTool {
  /// Current working directory
  cwd: Arc<RwLock<PathBuf>>,
}

impl GrepFilesTool {
  pub fn new(cwd: Arc<RwLock<PathBuf>>) -> Self {
    Self { cwd }
  }

  /// Search files for a pattern
  fn search_files(
    &self,
    cwd: &Path,
    pattern: &str,
    file_pattern: Option<&str>,
    case_insensitive: bool,
    max_matches: usize,
  ) -> Result<Vec<Match>, GrepFilesError> {
    // Build regex matcher
    let matcher = RegexMatcher::new_line_matcher(&pattern)
      .map_err(|e| GrepFilesError::InvalidPattern(e.to_string()))?;

    let mut matches = Vec::new();
    let mut files_searched = 0;

    // Walk the directory tree
    let walker = if let Some(pattern) = file_pattern {
      // TODO: Implement glob pattern filtering
      // For now, just walk all files
      WalkDir::new(cwd).max_depth(10).follow_links(false)
    } else {
      WalkDir::new(cwd).max_depth(10).follow_links(false)
    };

    for entry in walker {
      if matches.len() >= max_matches || files_searched >= MAX_FILES {
        break;
      }

      let entry = entry?;
      let path = entry.path();

      // Skip directories
      if !path.is_file() {
        continue;
      }

      // Skip binary files and large files
      if let Ok(metadata) = entry.metadata() {
        if metadata.len() > 10 * 1024 * 1024 {
          // Skip files > 10MB
          continue;
        }
      }

      files_searched += 1;

      // Search the file
      let mut searcher = SearcherBuilder::new()
        .binary_detection(BinaryDetection::quit(b'\x00'))
        .line_number(true)
        .build();

      let path_buf = path.to_path_buf();
      let result = searcher.search_path(
        &matcher,
        path,
        grep::searcher::sinks::UTF8(|line_number, line| {
          if matches.len() < max_matches {
            matches.push(Match {
              file: path_buf.clone(),
              line_number,
              line: line.trim_end().to_string(),
            });
            Ok(true)
          } else {
            Ok(false)
          }
        }),
      );

      // Ignore errors for individual files (e.g., permission denied)
      let _ = result;
    }

    Ok(matches)
  }
}

impl Tool for GrepFilesTool {
  const NAME: &'static str = "grep_files";

  type Args = GrepFilesArgs;
  type Output = String;
  type Error = GrepFilesError;

  async fn definition(&self, _prompt: String) -> ToolDefinition {
    ToolDefinition {
      name: "grep_files".to_string(),
      description: "Search for a pattern in files under the current working directory. Returns matching lines with file paths and line numbers.".to_string(),
      parameters: json!({
        "type": "object",
        "properties": {
          "pattern": {
            "type": "string",
            "description": "Pattern to search for (supports regex)"
          },
          "file_pattern": {
            "type": "string",
            "description": "File glob pattern to filter files (e.g., '*.rs', '*.py'). Optional."
          },
          "case_insensitive": {
            "type": "boolean",
            "description": "Whether to perform case-insensitive search. Default is false.",
            "default": false
          },
          "max_matches": {
            "type": "integer",
            "description": format!("Maximum number of matches to return (default: 50, max: {})", MAX_MATCHES),
            "minimum": 1,
            "maximum": MAX_MATCHES,
            "default": 50
          }
        },
        "required": ["pattern"]
      }),
    }
  }

  async fn call(&self, args: Self::Args) -> Result<Self::Output, Self::Error> {
    let cwd = self.cwd.read().map_err(|_| GrepFilesError::LockError)?;

    let max_matches = args.max_matches.unwrap_or(50).min(MAX_MATCHES);

    // Search files
    let matches = self.search_files(
      &cwd,
      &args.pattern,
      args.file_pattern.as_deref(),
      args.case_insensitive,
      max_matches,
    )?;

    // Format results
    if matches.is_empty() {
      return Ok(format!("No matches found for pattern: {}", args.pattern));
    }

    let mut result = format!("## Grep Results for '{}'\n\n", args.pattern);
    result.push_str(&format!("Found {} matches:\n\n", matches.len()));

    let mut current_file: Option<PathBuf> = None;

    for m in &matches {
      // Show file path when it changes
      if current_file.as_ref() != Some(&m.file) {
        if current_file.is_some() {
          result.push_str("\n");
        }
        result.push_str(&format!("### {}\n\n", m.file.display()));
        current_file = Some(m.file.clone());
      }

      result.push_str(&format!("{}:  {}\n", m.line_number, m.line));
    }

    if matches.len() >= max_matches {
      result.push_str(&format!(
        "\n*Showing first {} matches. There may be more.*",
        max_matches
      ));
    }

    Ok(result)
  }
}
